versiondb/tmdb/store.go

```diff
@@ -0,0 +1,157 @@
+package tmdb
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/crypto-org-chain/cronos/versiondb"
+	dbm "github.com/tendermint/tm-db"
+)
+
+var _ versiondb.VersionStore = (*Store)(nil)
+
+// Store implements `VersionStore`.
+type Store struct {
+	// latest key-value pairs
+	plainDB dbm.DB
+	// history bitmap index of keys
+	historyDB dbm.DB
+	// changesets of each blocks
+	changesetDB dbm.DB
+}
+
+func NewStore(plainDB, historyDB, changesetDB dbm.DB) *Store {
+	return &Store{plainDB, historyDB, changesetDB}
+}
+
+// PutAtVersion implements VersionStore interface
+// TODO reduce allocation within iterations.
+func (s *Store) PutAtVersion(version int64, changeSet []types.StoreKVPair) error {
+	plainBatch := s.plainDB.NewBatch()
+	defer plainBatch.Close() // nolint: errcheck
+	historyBatch := s.historyDB.NewBatch()
+	defer historyBatch.Close() // nolint: errcheck
+	changesetBatch := s.changesetDB.NewBatch()
+	defer changesetBatch.Close() // nolint: errcheck
+
+	for _, pair := range changeSet {
+		key := prependStoreKey(pair.StoreKey, pair.Key)
+
+		if version == 0 {
+			// genesis state is written into plain state directly
+			if pair.Delete {
+				return errors.New("can't delete at genesis")
+			} else {
+				if err := plainBatch.Set(key, pair.Value); err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		original, err := s.plainDB.Get(key)
+		if err != nil {
+			return err
+		}
+		if bytes.Equal(original, pair.Value) {
+			// do nothing if the value is not changed
+			continue
+		}
+
+		// write history index
+		if err := WriteHistoryIndex(s.historyDB, historyBatch, key, uint64(version)); err != nil {
+			return err
+		}
+
+		// write the old value to changeset
+		if len(original) > 0 {
+			changesetKey := append(sdk.Uint64ToBigEndian(uint64(version)), key...)
+			if err := changesetBatch.Set(changesetKey, original); err != nil {
+				return err
+			}
+		}
+
+		// write the new value to plain state
+		if pair.Delete {
+			if err := plainBatch.Delete(key); err != nil {
+				return err
+			}
+		} else {
+			if err := plainBatch.Set(key, pair.Value); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := changesetBatch.WriteSync(); err != nil {
+		return err
+	}
+	if err := historyBatch.WriteSync(); err != nil {
+		return err
+	}
+	return plainBatch.WriteSync()
```

Comment:
will we have dirty data if one `WriteSync` fails?

Reply:
It happens in end blocker, so it's before the chain state, so any error happens here, the block is not committed, when restart, the block will replay and the versiondb is written again for the same block.
The writes here should be idempotent, and it should be fine I think.

Reply:
oh if eventually consistent, maybe just only allow query after block get replayed?

Reply:
we can write a block number into plain db, that'd be the latest height, and we don't support query height higher than that.

Reply:
https://github.com/crypto-org-chain/cronos/pull/722/files#diff-aa83981ed9bef3f7410fa0d5177ca4934c6e4b3c7280205511237da9b3dc02abR92
you've updated the latest version, so only if` plainBatch.WriteSync()` succeeded and it will be updated. Right?

Reply:
yes